app/models/user.rb

```diff
@@ -117,102 +112,64 @@ def to_s
     name || "somebody"
   end
 
-  # Predictions used to be a tier 2 Patreon feature, so they're available to:
-  # 1. Gold users (the standard way of obtaining)
-  # 2. Continued patrons who used to have them at the lower price point (grandfathered)
-  # 3. Anyone paying at least the price of Gold via Patreon
+  # feature_grantors returns an array of objects that can give the user
+  # features, like a subscription, a grandfathered patreon subscription, or a
+  # trial. Every feature grantor must implement all `has_FEATURE?`-type methods
+  # below.
+  #
+  # If the feature grantor resolves to nil, it is assumed that it grants no
+  # features.
+  def feature_grantors
+    # TODO: Convert subscriptions from has-many to has-one
+    [patreon, subscriptions, trial].compact
```

Comment:
Is there any way to preload this so we're not making potentially 3 calls anytime we want to check features?

Reply:
I've tried, but it seems that Rails keeps optimizing the queries to be separate even with `includes`:

```rails
> User.includes(:patreon, :trial, :subscriptions).first.feature_grantors
  User Load (0.4ms)  SELECT "users".* FROM "users" ORDER BY "users"."id" ASC LIMIT $1  [["LIMIT", 1]]
  PatreonUser Load (0.3ms)  SELECT "patreon_users".* FROM "patreon_users" WHERE "patreon_users"."user_id" = $1  [["user_id", 1]]
  SubscriptionTrial Load (0.3ms)  SELECT "subscription_trials".* FROM "subscription_trials" WHERE "subscription_trials"."user_id" = $1  [["user_id", 1]]
  Subscription Load (0.4ms)  SELECT "subscriptions".* FROM "subscriptions" WHERE "subscriptions"."user_id" = $1  [["user_id", 1]]
```

I'm not super torn up about it, it's a maximum of 3 extra queries per page.

Reply:
Huh, that's weird.  I agree since it's a constant amount of queries (and not likely to grow) this isn't a big deal.